### ISSUE TITLE
[RFC] Code base improvements

### ISSUE THREAD

**Issue by D4N** [Exiv2/exiv2#218]
The past months have uncovered a number of (mostly) memory issues. These can be found quite easily via fuzzing which many volunteers have done and provided us with some valuable insight. We have been mostly patching these individually, but this does not solve the issues at the root and only on a case by case basis.

The past issues mostly fall in the following categories:
1. out of bounds read/writes due to bad range checks
2. bugs due to the usage of untrusted input, like an out of bounds read due to the length of an array being extracted from an input image
3. bugs due to integer overflows

I propose the following changes (further details below):
1. Replace `DataBuf` with a new container type that does not give raw access to the byte array but only via a range checked type.
2. Make the `getUShort()` functions return a quarantine type (e.g. `untrusted_t<>` or `quarantined_t<>`), that does allows for some basic handling (like arithmetics, storing) but is not convertible to the saved type. It would only be convertible to the underlying type via an explicit check.
3. Expand & optimize the overflow checked operations in `src/safe_op.hpp`.
4. Start replacing all the standard types like `long`, `int`, `long long` with fixed size (`int64_t`, `uint32_t` etc.) or other special purpose types (`size_t`, `uintptr_t`), where applicable.

### New Container
Originally, I wanted to use `std::vector` & `std::array` as the new `DataBuf`, but both have a rather nasty drawback: you cannot really conveniently do this:
``` C++
const unsigned short val = getUShort(buf.data_ + 4, byteOrder);
```
It is possible to do this with iterators, but iterators are not range checked. So that would retain the same disadvantage, while making the code less readable. We could use the "standard" approach for this, by changing the API in the following way:
``` C++
usigned short getUshort(const iterator start, const iterator end, byteOrder);
```
Personally, I think that's not a good API, as it is needlessly verbose and requires more error checking (nothing prevents you from passing `end < start`) and unnecessarily more work by the programmer. Also, it's more complicated then before.

Therefore, I think we should either implement our own container or a range checked iterator for `std::vector`. Although the custom container is more work, it could be beneficial, as the API could be more uniform.

I was thinking about using a container and a slice type, where the container is something like `std::vector` and the slice represents a part of the vector (the slice would also only hold a pointer/reference to the container). Both would provide an `at(size_t index)` function, which would either return a reference to the requested element at `index` or throw an exception.

I have experimented a bit with this and wanted to make the container as immutable as possible (i.e. length is constant & set on construction). Unfortunately this makes it impossible to use it as a class member where the length is decided later or to re-use the container with a different length, at least directly. One could achieve this indirectly by storing the container on the heap in a `unique_ptr`.

Maybe a wrapper around `std::vector` would be way to go?

### Untrusted input
The encapsulating type for untrusted values could be something like this:
``` C++
template <typename T>
struct untrusted_t {
    template <typename F, int err_code>
    T get_value(F is_valid) {
        if (!is_valid(val_)) {
            throw Error(err_code);
        }
        return val_;
    }
  private:
    const T val_;
};
```
The idea is, that to actually use the saved value, you have to provide the `get_value` function with a custom check function that forces you to verify that the input is ok for further usage. With C++11 lambdas, this can be achieved like this:
``` C++
const auto extracted_length = getULong(...);
const unsigned long length = extracted_length.get_value(
    [&io](unsigned int val) { return val < io.size(); }
);
```
In a nutshell this should prevent one from just taking input from a file before doing a sanity check while also not requiring to write the `if (cond) { throw ... ; } else { val = ... }`  boilerplate. However it might be possible that we have to implement a raw getter for data where no check is possible, like the path which is saved in the image metadata.
Whether this type would require arithmetic operators (which would always check for overflows) is debatable. We could of course simply save most of the metadata as `untrusted_t` thereby indicating downstream library users, that they should really check the values before using them. But then we'll break a lot of code and everybody will hate us.

Any comments & suggestions are welcome.

**Comment by clanmills** [comment]
Dan

Thanks for thinking about this and making a proposal.  I normally like to encourage everybody.  I strongly support that we review/discontinue the use of 'long', 'int' and adopt your suggestions.

However, when I read your proposal about using std::vector, I'm wondering what problem is being solved by your proposal.  For sure you're adding complexity by passing iterators to functions.  Can we agree to define the design goal is "we should never access the raw bytes in a DataBuf"?

We can address boiler-plate such as `if (cond) { throw ... ; } else { val = ... }` using a function such as `val = io. getUShort(int32_t offset=0)`.   We can use such an API to read from a DataBuf using the MemIo object to wrap the memory as a stream, something like:  `MemIo(dataBuf).getUShort()`

My preference is to retain DataBuf and remove all "raw" byte access in the code.  If we change the declaration of pData_ to **_private_**, the compiler will reveal the code that should be modified.

**Comment by D4N** [comment]
I am not convinced to use std::vector, although it has its advantages. First we wouldn't need to ship a custom container ourselves and second the "auxiliary functions" (something like `getUShort`) could be easier used by others if they used STL containers instead of a custom one. I was simply calling the idea a "new container" so that we would have the possibility to deprecate DataBuf and remove it slowly from the code base. Shaping DataBuf to our needs is fine by me, too.

I agree that we should completely disallow direct access to the array. Also, we should alter the behavior of the copy constructor & assignment operator for C++11 (but that belongs to the respective issue).

I like your idea of including `offset` into the signature of the `getUShort` class of functions. Hopefully that will be enough and we do not need something like slices. I however wouldn't wrap the DataBuf in a MemIo object, as that would only incur virtual function calls in every call to `getType()`.

**Comment by clanmills** [comment]
@D4N It sounds as though we agree to:
1) Review/discontinue the use of 'long', 'int' and adopt your suggestions
2) Modify class DataBuf to deny direct data access to "raw" bytes
3) Modify class BasicIo to add getUShort() and other data access functions
4) Remove static API functions such as `EXIV2API Exiv2::getUShort()` http://www.exiv2.org/doc/namespaceExiv2.html
5) Expand & optimize the overflow checked operations in `src/safe_op.hpp`

I wouldn't worry about the performance of virtual functions.  It's possible may never require MemIo(dataBuf).getUShort().  The point I was making is that we don't need to retain `EXIV2API Exiv2::getUShort()` to use our 'safe' code on bytes in memory somewhere.

The C++ compiler knows the index of a virtual function in the vtable and generates code such as `*(vtable_[index_])()`  For a "regular" class function, the compiler/linker knows the address of the function and generates code such as `*(address)(...)`.  The difference in performance in almost nothing.
